torch/utils/flop_counter.py

```diff
@@ -0,0 +1,306 @@
+    m, k = a_shape
+    k2, n = b_shape
+    assert k == k2
+    # NB(chilli): Should be 2 * k - 1 technically for FLOPs.
+    return m * n * 2 * k
+
+def addmm_flop(input_shapes: List[Any], _=None) -> int:
+    """
+    Count flops for addmm
+    """
+    assert len(input_shapes) == 3
+    _, a_shape, b_shape = input_shapes
+    mm_flops = mm_flop([a_shape, b_shape])
+    batch_size, _ = a_shape
+    _, output_dim = b_shape
+    return mm_flops + batch_size * output_dim
+
+def bmm_flop(input_shapes: List[Any], _=None) -> int:
+    """
+    Count flops for the bmm operation.
+    """
+    # Inputs should be a list of length 2.
+    # Inputs contains the shapes of two tensor.
+    assert len(input_shapes) == 2, len(input_shapes)
+    a_shape, b_shape = input_shapes
+    b, m, k = a_shape
+    b2, k2, n = b_shape
+    assert b == b2
+    assert k == k2
+    # NB(chilli): Should be 2 * k - 1 technically for FLOPs.
+    flop = b * m * n * 2 * k
+    return flop
+
+def baddbmm_flop(input_shapes: List[Any], _=None) -> int:
+    """
+    Count flops for the baddbmm operation.
+    """
+    # Inputs should be a list of length 3.
+    # Inputs contains the shapes of three tensors.
+    assert len(input_shapes) == 3, len(input_shapes)
+    _, a_shape, b_shape = input_shapes
+    bmm_flops = bmm_flop([a_shape, b_shape])
+    b, m, _ = a_shape
+    _, _, n = b_shape
+    flop = bmm_flops + b * m * n
+    return flop
+
+
+def conv_flop_count(
+    x_shape: List[int],
+    w_shape: List[int],
+    out_shape: List[int],
+    transposed: bool = False,
+) -> int:
+    """
+    Count flops for convolution. Note only multiplication is
+    counted. Computation for addition and bias is ignored.
+    Flops for a transposed convolution are calculated as
+    flops = (x_shape[2:] * prod(w_shape) * batch_size).
+    Args:
+        x_shape (list(int)): The input shape before convolution.
+        w_shape (list(int)): The filter shape.
+        out_shape (list(int)): The output shape after convolution.
+        transposed (bool): is the convolution transposed
+    Returns:
+        int: the number of flops
+    """
+    batch_size = x_shape[0]
+    conv_shape = (x_shape if transposed else out_shape)[2:]
+    c_out, c_in, *dims = w_shape
+
+    # NB(chilli): I don't think this properly accounts for padding :think:
+    # NB(chilli): Should be 2 * c_in - 1 technically for FLOPs.
+    flop = batch_size * prod(conv_shape) * c_out * prod(dims) * 2 * c_in
+    return flop
+
+def conv_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
+    """
+    Count flops for convolution.
+    """
+    x_shape, w_shape = input_shapes[:2]
+    out_shape = output_shapes[0]
+    transposed = input_shapes[6]
+
+    return conv_flop_count(x_shape, w_shape, out_shape, transposed=transposed)
+
+def transpose_shape(shape):
+    return [shape[1], shape[0]] + list(shape[2:])
+
+def conv_backward_flop(input_shapes: List[Any], output_shapes: List[Any]) -> int:
```

Comment:
Should you just pass in the args/kwargs to these functions? That will simplify the logic in here quite a bit no?

Reply:
Hmm... maybe.